### PR TITLE
Handle redirects in apiserver proxy handler

### PR DESCRIPTION
**Pull request by timstclair**
Overview:
1. Peek at the HTTP response from the proxied backend
2. If it is a redirect response (302/3), redo the request to the redirect location
3. If it's not a redirect, forward the response to the client and then set up the proxy as before

This change is required for implementing streaming requests in the Container Runtime Interface (CRI). See [design](https://docs.google.com/document/d/1OE_QoInPlVCK9rMAx9aybRmgFiVjHpJCHI9LrfdNM_s/edit).

For https://github.com/kubernetes/kubernetes/issues/29579

/cc @yujuhong 

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34987)

### CODE REVIEW

**Comment by ncdc** [comment]
I'll try to take a look at this later today or tomorrow. Also cc @liggitt

**Comment by liggitt** [comment]
need to take a closer look, but at the very least, this should be opt-in... not all proxy handling should follow redirects server-side

**Comment by timstclair** [comment]
> at the very least, this should be opt-in... not all proxy handling should follow redirects server-side

Makes sense. Here's a couple options:

1. Make it configurable in the proxy, e.g. configure the proxy for exec/attach/portforward to follow redirects
2.  Make it configurable per-request/response, e.g. a header specifies whether the proxy should follow the redirect
3. I suppose the traditional solution from a reverse proxy perspective would be to not handle the redirect at all, and just rewrite the location header with the external address (i.e. <apiserver>/proxy/...)

I'm leaning towards option (1), since the apiserver can expect the redirect in that case. Option (3) could work, but it would add an extra hop to the connection (might be a non-issue?), and the apiserver would need to know if redirect location was a pod or a node.

Am I missing anything?

/cc @yujuhong

**Comment by liggitt** [comment]
option 1, feature-gated with CRI-enablement, seems like the best approach. still want to dig into the mechanism (and ideally factor it out of the main flow a little better)

**Comment by yujuhong** [comment]
> option 1, feature-gated with CRI-enablement, seems like the best approach. still want to dig into the mechanism (and ideally factor it out of the main flow a little better)

Option 1 sounds good to me, too.

I don't think we need to feature gate this in the apiserver. CRI affects how kubelet integrates with the runtime, and should be feature gated there (on the kubelet side). This should not concern apiserver, and apiserver should not receive redirects (for exec/attach/portforward) at all if the feature is not enabled in kubelet.

**Comment by liggitt** [comment]
> I don't think we need to feature gate this in the apiserver

sniffing responses and redirecting is a pretty invasive change (as evidenced by this PR)... it should be gated

**Comment by timstclair** [comment]
I refactored this based on the discussion here. Redirect interception is now only enabled for exec/attach/port-forward requests, and can be disabled with a feature gate (`StreamingProxyRedirects`).

File: pkg/registry/generic/rest/proxy.go [hunk 1]
```diff
@@ -213,21 +324,20 @@ func (h *UpgradeAwareProxyHandler) defaultProxyTransport(url *url.URL, internalT
...
+func (rt *corsRemovingTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	resp, err := rt.RoundTripper.RoundTrip(req)
 	if err != nil {
 		return nil, err
 	}
 	removeCORSHeaders(resp)
 	return resp, nil
 }
 
-var _ = net.RoundTripperWrapper(&corsRemovingTransport{})
```
Comment (liggitt) [inline pkg/registry/generic/rest/proxy.go]: don't drop this, let the compiler help us

File: pkg/util/config/feature_gate.go [hunk 2]
```diff
@@ -42,17 +42,19 @@ const (
...
+	streamingProxyRedirects   = "StreamingProxyRedirects"
 )
 
 var (
 	// Default values for recorded features.  Every new feature gate should be
 	// represented here.
 	knownFeatures = map[string]featureSpec{
 		allAlphaGate:              {false, alpha},
-		externalTrafficLocalOnly:  {false, alpha},
+		externalTrafficLocalOnly:  {true, beta},
```
Comment (liggitt) [inline pkg/util/config/feature_gate.go]: unrelated?

File: pkg/registry/generic/rest/proxy.go [hunk 3]
```diff
@@ -130,33 +137,35 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
 	if !httpstream.IsUpgradeRequest(req) {
 		return false
 	}
-
-	backendConn, err := proxy.DialURL(h.Location, h.Transport)
-	if err != nil {
+	if err := h.handleUpgrade(w, req); err != nil {
```
Comment (liggitt) [inline pkg/registry/generic/rest/proxy.go]: This change moves the hijack and close of the request connection before we get a chance to respond to errors dialing the backend. That means we would no longer send API error responses in error cases.

Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: Good catch! It turns out this is an existing issue, since the ErrorResponder fails once the connection is hijacked. I added a test case for this, and fixed it by writing the errors to the hijacked connection directly.

File: pkg/registry/generic/rest/proxy.go (hunk 1 above)
Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: The compiler already enforces this since we use it in a `net.RoundTripperWrapper` context, and the `net` package name now conflicts with the stdlib `net`. If you feel strongly I can alias the package name and add it back.

File: pkg/util/config/feature_gate.go (hunk 2 above)
Comment (timstclair) [inline pkg/util/config/feature_gate.go]: Just something weird about the diffing... I rebased on HEAD and this resolved.

**Review by timstclair — COMMENTED** [review]
Thanks for reviewing! Responded to all comments.

**Comment by liggitt** [comment]
> It turns out this is an existing issue, since the ErrorResponder fails once the connection is hijacked. I added a test case for this, and fixed it by writing the errors to the hijacked connection directly.

The previous flow was intentional... let the error responder report errors during backend establishment, then hijack. The error responder can't write once we've written content to the connection (headers are already committed, and a well formed API error JSON blob after random other content isn't usable)

File: pkg/registry/generic/rest/proxy.go [hunk 4]
```diff
@@ -240,3 +351,25 @@ func removeCORSHeaders(resp *http.Response) {
...
+
+type hijackedErrorResponder struct {
+	// The hijacked client connection to respond over.
+	conn net.Conn
+}
+
+func (r *hijackedErrorResponder) Error(err error) {
+	header := http.Header{}
+	header.Set("Content-Type", "text/plain")
+	body := bytes.NewBufferString(err.Error())
```
Comment (liggitt) [inline pkg/registry/generic/rest/proxy.go]: Wasn't this JSON before (or maybe negotiated?)

**Comment by timstclair** [comment]
> The previous flow was intentional... let the error responder report errors during backend establishment, then hijack.

Hmm, I see. I'll refactor it a bit tomorrow wait until after the BE connection is established to hijack.

There was still an error in the old flow though, because the Responder was invoked after the connection was hijacked ([here](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/generic/rest/proxy.go#L150)). I guess the hijacking should just be moved so that it's the last step before the proxy is established?

**Comment by liggitt** [comment]
Yeah, the request construction block might be able to be moved above the hijack. Yeah, hijack at the last possible moment

**Comment by timstclair** [comment]
Done. PTAL.

File: pkg/registry/generic/rest/proxy.go (hunk 1 above)
Comment (liggitt) [inline pkg/registry/generic/rest/proxy.go]: I'm not seeing the type assertion elsewhere, I'd like to keep it

**Comment by liggitt** [comment]
Will finish up review tomorrow

**Comment by yujuhong** [comment]
@liggitt @ncdc a friendly ping since code freeze is imminent!

File: pkg/registry/generic/rest/proxy.go [hunk 5]
```diff
@@ -131,32 +138,42 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
...
+		err         error
+	)
+	if h.InterceptRedirects && utilconfig.DefaultFeatureGate.StreamingProxyRedirects() {
+		backendConn, rawResponse, err = h.connectBackendWithRedirects(req)
+	} else {
+		backendConn, err = h.connectBackend(req, h.Location)
+	}
 	if err != nil {
-		h.Responder.Error(err)
+		h.Responder.Error(fmt.Errorf("error dialing backend: %v", err))
```
Comment (ncdc) [inline pkg/registry/generic/rest/proxy.go]: Depending on what causes `connectBackend`/`connectBackendWithRedirects` to return an error, you may end up writing `error dialing backend: error dialing backend: ...`. It would be nice to avoid that if possible.

File: pkg/registry/generic/rest/proxy.go [hunk 6]
```diff
@@ -131,32 +138,42 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
...
 	}
-	newReq.Header = req.Header
+	defer requestHijackedConn.Close()
 
-	if err = newReq.Write(backendConn); err != nil {
-		h.Responder.Error(err)
-		return true
+	// Forward raw response bytes back to client.
+	if _, err = requestHijackedConn.Write(rawResponse); err != nil {
+		glog.Errorf("Error proxying response from backend to client: %v", err)
```
Comment (ncdc) [inline pkg/registry/generic/rest/proxy.go]: `utilruntime.HandleError(fmt.Errorf("error proxying response...`

File: pkg/registry/generic/rest/proxy.go [hunk 7]
```diff
@@ -192,6 +209,104 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
...
+		return nil, err
+	}
+	beReq.Header = req.Header
+
+	conn, err = proxy.DialURL(location, h.Transport)
+	if err != nil {
+		return conn, fmt.Errorf("error dialing backend: %v", err)
+	}
+
+	if err = beReq.Write(conn); err != nil {
```
Comment (liggitt) [inline pkg/registry/generic/rest/proxy.go]: doesn't this consume the original `req.Body`?

File: pkg/registry/generic/rest/proxy.go [hunk 8]
```diff
@@ -192,6 +209,104 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
...
+		}
+	}()
+
+redirectLoop:
+	for redirects := 0; ; redirects++ {
+		if redirects == maxRedirects {
+			return nil, nil, fmt.Errorf("too many redirects (%d)", redirects)
+		}
+
+		intermediateConn, err = h.connectBackend(req, location)
```
Comment (liggitt) [inline pkg/registry/generic/rest/proxy.go]: doesn't calling `connectBackend` consume the req.Body the first time? won't that cause failures in subsequent calls?

File: pkg/registry/generic/rest/proxy.go (hunk 5 above)
Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: Done.

File: pkg/registry/generic/rest/proxy.go (hunk 6 above)
Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: Done (What does this give us? Should I update the other error logging in the go routines below too?)

File: pkg/registry/generic/rest/proxy.go (hunk 8 above)
Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: Hmm, I think so, but this is also consistent with what `http.Client.Do` does. That implementation changes the redirected requests to GET (from POST) requests though, and I bet this is why. This shouldn't affect our usage, since these requests don't have bodies anyway. What do you think the best way to deal with it is?

File: pkg/registry/generic/rest/proxy.go (hunk 1 above)
Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: done

**Comment by ncdc** [comment]
I would feel more comfortable with something like this at the beginning of the dev cycle as opposed to just before code freeze. We may need to be prepared to revert and redo if things start behaving oddly.

**Review by timstclair — COMMENTED** [review]
Thanks, addressed comments. Open question about handling request body in redirects.

> I would feel more comfortable with something like this at the beginning of the dev cycle as opposed to just before code freeze. We may need to be prepared to revert and redo if things start behaving oddly.

Ack. There is a feature flag if need be, and this code path shouldn't be exercised normally anyway. Also, now is just before feature freeze, we still have a few more weeks of bug fixing, testing and stabilization.

**Comment by timstclair** [comment]
Friendly ping. This is getting down to the wire. If you still have concerns over this, could we schedule a hangout to discuss it?

File: pkg/registry/generic/rest/proxy.go (hunk 8 above)
Comment (ncdc) [inline pkg/registry/generic/rest/proxy.go]: Thinking about this a bit more... If we decide to move to HTTP/2 for streaming requests, and if there are no changes to go's HTTP/2 library, then the only means we'll have of streaming data over the wire will be via the request and response bodies (we'd need to implementing muxing on top). Which would mean that when the client has data it wants to send to the server, the original request body needs to be preserved and available. I'm not sure how that would work... Also not sure it needs to stop this from going in.

File: pkg/registry/generic/rest/proxy.go [hunk 9]
```diff
@@ -131,32 +140,42 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
...
 		return true
 	}
-	newReq.Header = req.Header
+	defer requestHijackedConn.Close()
 
-	if err = newReq.Write(backendConn); err != nil {
-		h.Responder.Error(err)
-		return true
+	// Forward raw response bytes back to client.
+	if _, err = requestHijackedConn.Write(rawResponse); err != nil {
```
Comment (ncdc) [inline pkg/registry/generic/rest/proxy.go]: I guess it's ok to do a 0-byte Write in the event we aren't following a redirect?

**Review by ncdc — COMMENTED** [review]
If the tests are happy, I'm ok with this going in. But this definitely has the possibility to be the source of some weird or flakey behavior w.r.t exec / attach / port forward.

File: pkg/registry/generic/rest/proxy.go (hunk 7 above)
Comment (sttts) [inline pkg/registry/generic/rest/proxy.go]: it proxies the orignals request's body to the backend. So that's the intention, isn't it?

Comment (liggitt) [inline pkg/registry/generic/rest/proxy.go]: yes, but connectBackend() is called repeatedly with the same request in cases where the backend returns a redirect

File: pkg/registry/generic/rest/proxy.go [hunk 10]
```diff
@@ -192,6 +211,104 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
...
+		resp, err := http.ReadResponse(respReader, req)
+		if err != nil {
+			// Unable to read the backend response; let the client handle it.
+			glog.Warningf("Error reading backend response: %v", err)
+			break redirectLoop
+		}
+		resp.Body.Close() // Unused.
+
+		switch resp.StatusCode {
+		case http.StatusFound, http.StatusSeeOther:
```
Comment (sttts) [inline pkg/registry/generic/rest/proxy.go]: 303 => POST becomes GET. Do we want that?

File: pkg/registry/generic/rest/proxy.go (hunk 8 above)
Comment (sttts) [inline pkg/registry/generic/rest/proxy.go]: Can't we buffer a limited amount of the body and re-send it again and again? Like n<=1000 bytes. Every server should know what to do after those n bytes. If it asks for more, we continue to assume that there is no redirect.

File: pkg/registry/generic/rest/proxy.go (hunk 10 above)
Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: This was originally copied from the http package, but I removed it to narrow the scope (since we're serving a very specific use case here).

File: pkg/registry/generic/rest/proxy.go (hunk 9 above)
Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: I think so, but decided to check the len to be safe.

**Comment by timstclair** [comment]
Addressed outstanding issues. Also, switched to default-off alpha based on the number of corner cases and concerns.

File: pkg/registry/generic/rest/proxy.go [hunk 11]
```diff
@@ -192,6 +213,112 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
...
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// Peek at the backend response.
+		rawResponse.Reset()
+		respReader := bufio.NewReader(io.TeeReader(
+			io.LimitReader(intermediateConn, maxResponseSize), // Don't read more than maxResponseSize bytes.
+			rawResponse)) // Save the raw response.
+		resp, err := http.ReadResponse(respReader, req)
```
Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: If the maxResponseSize is hit, `ReadResponse` would return an error, which would cause the whole response to be sent to the client (i.e. no redirects followed). We don't read the body, so if it lands in the middle of the body it would be a non-issue.

Comment (liggitt) [inline pkg/registry/generic/rest/proxy.go]: looks like it reads the HTTP protocol line, reads the headers, and wraps the remainder in a `Body` reader in the returned response, so the only issue would be if maxResponseSize fell in the headers portion of the response

File: pkg/registry/generic/rest/proxy.go [hunk 12]
```diff
@@ -192,6 +213,112 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
...
+
+	beReq, err := http.NewRequest(method, location.String(), body)
+	if err != nil {
+		return nil, err
+	}
+	beReq.Header = header
+
+	conn, err = proxy.DialURL(location, h.Transport)
+	if err != nil {
+		return conn, fmt.Errorf("error dialing backend: %v", err)
```
Comment (liggitt) [inline pkg/registry/generic/rest/proxy.go]: return nil in error scenarios? otherwise we try to double close it, right?

File: pkg/registry/generic/rest/proxy.go (hunk 11 above)
Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: Hmm, I'm not positive, but I think if the maxResponseSize fell at just the right point between headers, it could look like a complete response and drop some headers. The worst case scenario would be that it was a redirect response (302), and the "Location" header was dropped. In this case, an error would be returned to the client instead: `302 response missing Location header`.

I think options here are to ignore this case, or always forward the raw response to the client in the case of an error following a redirect.

File: pkg/registry/generic/rest/proxy.go (hunk 12 above)
Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: Good catch. Done.

File: pkg/registry/generic/rest/proxy.go (hunk 11 above)
Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: I issued a bunch of exec/attach/portforward calls, and logged the length of the responses. They were all < 200 bytes. I arbitrarily picked 1024 as a permissive upper bound that should still have minimal impact on memory if it's hit.

Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: From out of band conversation: this imposes an artificial limit on the length of the exec command, since it is included in the Location header.

Comment (liggitt) [inline pkg/registry/generic/rest/proxy.go]: talked offline, once the kubelet starts redirecting, if it includes the original query (containing exec/attach command and params) that's arbitrary-length user-data. That probably rules out that method. This is fine for alpha, but might require rethinking as the kubelet pieces get built out.

**Comment by liggitt** [comment]
this LGTM, go ahead and squash

**Comment by liggitt** [comment]
hmmm
```
go test -v k8s.io/kubernetes/pkg/registry/generic/rest -run TestProxyUpgrade$
proxy_test.go:437: http with redirect: websocket dial err: websocket.Dial ws://127.0.0.1:60163/some/path: bad status
```

**Comment by timstclair** [comment]
Oops, I forgot to flip the feature flag back on for the test.

**Comment by timstclair** [comment]
Filed https://github.com/kubernetes/kubernetes/issues/36187 to track follow up work.

**Comment by timstclair** [comment]
Squashed.

**Comment by timstclair** [comment]
@k8s-bot unit test this https://github.com/kubernetes/kubernetes/issues/32455

**Comment by timstclair** [comment]
Reran `hack/update-bazel.sh`

File: pkg/registry/generic/rest/proxy.go [hunk 13]
```diff
@@ -192,6 +213,113 @@ func (h *UpgradeAwareProxyHandler) tryUpgrade(w http.ResponseWriter, req *http.R
...
+	)
+	defer func() {
+		if intermediateConn != nil {
+			intermediateConn.Close()
+		}
+	}()
+
+redirectLoop:
+	for redirects := 0; ; redirects++ {
+		if redirects == maxRedirects {
```
Comment (sttts) [inline pkg/registry/generic/rest/proxy.go]: Off by one. With `maxRedirects==0` we should at least connect once.

Comment (liggitt) [inline pkg/registry/generic/rest/proxy.go]: follow up for this is fine

Comment (timstclair) [inline pkg/registry/generic/rest/proxy.go]: Thanks, fixed.

**Comment by timstclair** [comment]
Fixed off-by-one. Reapplying LGTM.

**Comment by yujuhong** [comment]
Marking p2 to ensure PR dependency is respected.

**Comment by timstclair** [comment]
@k8s-bot cvm gce e2e test this https://github.com/kubernetes/kubernetes/issues/33380
